master/internal/experiment_filter.go

```diff
@@ -132,6 +132,111 @@ func expColumnNameToSQL(columnName string) (string, error) {
 	return col, nil
 }
 
+func runColumnNameToSQL(columnName string) (string, error) {
+	// To prevent SQL injection this function should never
+	// return a user generated field name
+
+	filterExperimentColMap := map[string]string{
+		"id":                    "r.id",
+		"experimentDescription": "e.config->>'description'",
+		"experimentName":        "e.config->>'name'",
+		"tags":                  "e.config->>'labels'",
+		"searcherType":          "e.config->'searcher'->>'name'",
+		"searcherMetric":        "e.config->'searcher'->>'metric'",
+		"startTime":             "r.start_time",
+		"endTime":               "r.end_time",
+		"duration":              "extract(epoch FROM coalesce(r.end_time, now()) - r.start_time)",
+		"state":                 "r.state",
+		"experimentProgress":    "ROUND(COALESCE(progress, 0) * 100)::INTEGER", // multiply by 100 for percent
+		"user":                  "e.owner_id",
+		"forkedFrom":            "e.parent_id",
+		"resourcePool":          "e.config->'resources'->>'resource_pool'",
+		"projectId":             "r.project_id",
+		"checkpointSize":        "e.checkpoint_size",
+		"checkpointCount":       "e.checkpoint_count",
+		"searcherMetricsVal":    "r.searcher_metric_value",
+		"externalExperimentId":  "e.external_experiment_id",
+		"externalTrialId":       "r.external_run_id",
+		"experimentId":          "e.id",
+	}
+	var exists bool
+	col, exists := filterExperimentColMap[columnName]
+	if !exists {
+		return "", fmt.Errorf("invalid run column %s", columnName)
+	}
+	return col, nil
+}
+
+func runHpToSQL(c string, filterColumnType *string, filterValue *interface{},
+	op *operator, q *bun.SelectQuery,
+	fc *filterConjunction,
+) (*bun.SelectQuery, error) {
+	queryColumnType := projectv1.ColumnType_COLUMN_TYPE_UNSPECIFIED.String()
+	var o operator
+	var queryValue interface{}
+	if filterValue == nil && op != nil && *op != empty && *op != notEmpty {
+		return nil, fmt.Errorf("hyperparameter field defined without value and without a valid operator")
+	}
+	o = *op
+	if o != empty && o != notEmpty {
+		queryValue = *filterValue
+	}
+	if filterColumnType != nil {
+		queryColumnType = *filterColumnType
+	}
+	hp := strings.Split(strings.TrimPrefix(c, "hp."), ".")
+	for i := 0; i < len(hp); i++ {
+		// for last element
+		if i == len(hp)-1 {
+			hp[i] = fmt.Sprintf(`>'%s'`, hp[i])
+		} else {
+			hp[i] = fmt.Sprintf(`'%s'`, hp[i])
+		}
+	}
+	hpQuery := strings.Join(hp, "->")
+	oSQL, err := o.toSQL()
+	if err != nil {
+		return nil, err
+	}
+	var queryArgs []interface{}
+	var queryString string
+	switch o {
+	case empty:
+		queryString = fmt.Sprintf(`r.hparams->%s IS NULL`, hpQuery)
+	case notEmpty:
+		queryString = fmt.Sprintf(`r.hparams->%s IS NOT NULL`, hpQuery)
+	case contains:
+		queryArgs = append(queryArgs, queryValue)
+		if queryColumnType == projectv1.ColumnType_COLUMN_TYPE_TEXT.String() ||
```

Comment:
nit: it feels a more natural to cast when ```queryColumnType == numeric``` rather than casting when it isn't a string/date